src/Tasks/Microsoft.Common.CurrentVersion.targets

```diff
@@ -720,6 +720,110 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     </EmbeddedResource>
   </ItemDefinitionGroup>
 
+  <!--
+    ========================
+    DeterministicSourcePaths
+    ========================
+    
+    Unless specified otherwise enable deterministic source root when building deterministically on CI server, but not for local builds.
+    In order for the debugger to find source files when debugging a locally built binary the PDB must contain original, unmapped local paths.
+  -->
+  <PropertyGroup>
+    <DeterministicSourcePaths Condition="'$(DeterministicSourcePaths)' == '' and '$(Deterministic)' == 'true' and '$(ContinuousIntegrationBuild)' != ''">true</DeterministicSourcePaths>
+  </PropertyGroup>
+
+  <!--
+    ==========
+    SourceRoot
+    ==========
+    
+    All source files of the project are expected to be located under one of the directories specified by SourceRoot item group.
+    This targets collects all SourceRoots from various sources (InitializeSourceRootDependsOn targets).
+    
+    If DeterministicSourcePaths is true it calculates deterministic mapped paths that can be used by CoreCompile targets
+    to initialize deterministic path mapping specific to the compiler.
+    
+    Include the final backslash in the path of each SourceRoot item.
+    
+    Extensibility: Add target to InitializeSourceRootDependsOn that populates SourceRoot item group.
+  -->
+
+  <Target Name="InitializeSourceRoot"
+          DependsOnTargets="$(InitializeSourceRootDependsOn)">
+    <ItemGroup>
+      <_InvalidSourceRoot Include="@(SourceRoot)" Condition="!HasTrailingSlash('%(SourceRoot.Identity)')" />
+    </ItemGroup>
+
+    <Error Text="SourceRoot paths are required to end with a slash or backslash: @(_InvalidSourceRoot, ', ')"
+           Condition="'@(_InvalidSourceRoot)' != ''" />
```

Comment:
Why can't we just add them? That is what common targets normally does for this.

Reply:
Do we know which separator to add on Windows (`\` or `/`)? The source root path needs to end with exactly the same separator that the full paths of source files that flow into the compiler use. 
Is it possible that the compiler will see a source path with `/` in it on Windows?

Reply:
There's some magic that changes backslash to slash for ToolTask invocation on Unix

Take a deep breath:

https://github.com/Microsoft/msbuild/blob/cc9cc0c54454d427b12d73255b950908894ed10e/src/Utilities/ToolTask.cs#L1725

So I think appending \ will give the compiler \ on Windows and / on Unix.

Reply:
I think it is possible for the compiler to see / on Windows, but not \ on Unix.

Reply:
Why can't shouldn't the compiler understand that there's no difference between / and \ on Windows?

Reply:
Source paths in the PDB are matched to the patterns in SourceLink by simple prefix match when interpreted by the debugger.

I don't think requiring the path to end with slash is an issue. SourceRoot items will usually be generated by another build task. Writing SourceRoots manually should be very rare.

Reply:
I don't think it's much of an issue. I asked because the append-trailing-slash-if-not-already-there idiom is pervasive in common targets. If there's a good reason to do it differently here, that's fine. Might be worth a comment.

Reply:
Will add a comment.